File: apps/dsa-web/src/types/btcTraining.ts
export type BtcTrainingArchitecture = 'patchtst' | 'itransformer' | 'fusion' | 'all';

export interface BtcTrainingConfig {
  architectures: string[];
  defaultSeeds: number[];
  minSeeds: number;
  minEpochs: number;
  featureCount: number;
  featureColumns: string[];
  researchOnly: boolean;
  promotionEligible: boolean;
}

export interface BtcTrainingRunRequest {
  architecture: BtcTrainingArchitecture;
  epochs: number;
  seeds: number[];
  sequenceLength?: number;
  folds?: number;
  minTrainSamples?: number;
  validationSamples?: number;
  purgeSamples?: number;
  ablationFeatures?: string[];
}

export interface BtcTrainingTaskAccepted {
  taskId: string;
  status: string;
  message?: string;
  protocol: {
    sameLabels: boolean;
    sameValidationWindow: boolean;
    seedCount: number;
    epochs: number;
    architectures: string[];
    featureCount?: number;
    researchOnly: boolean;
    promotionEligible: boolean;
  };
}

export interface BtcTrainingRunResult {
  architecture: string;
  seed: number;
  featureSet: string;
  removedFeature?: string | null;
  featureCount: number;
  dataQuality?: string;
  oofPredictionCount: number;
  labelSignature?: string;
  validationWindowSignature?: string;
  evaluations?: Record<string, {
    foldCount?: number;
    samples?: number;
    directionAccuracy?: number | null;
    returnMae?: number | null;
    pearsonIc?: number | null;
    spearmanIc?: number | null;
  }>;
}

export interface BtcTrainingResearchResult {
  mode: string;
  researchOnly: boolean;
  participatesInDecision: boolean;
  eligibleForPromotion: boolean;
  promotionEligible: boolean;
  artifactPath?: string;
  oofArtifactPath?: string;
  artifactRole?: string;
  oofPredictionCount: number;
  protocol: {
    sameLabels: boolean;
    sameValidationWindow: boolean;
    labelSignature?: string;
    validationWindowSignature?: string;
    ablation: string;
    seedCount: number;
    seeds: number[];
    epochs: number;
    architectures: string[];
    featureCount: number;
    featureColumns: string[];
    ablationFeatures: string[];
  };
  summary: Record<string, Record<string, {
    runs: number;
    availableRuns: number;
    meanDirectionAccuracy?: number | null;
    meanReturnMae?: number | null;
    meanPearsonIc?: number | null;
    meanSpearmanIc?: number | null;
  }> >;
  runs: BtcTrainingRunResult[];
}

export interface BtcTrainingTaskStatus {
  taskId: string;
  status: string;
  progress: number;
  message?: string;
  result?: BtcTrainingResearchResult;
  error?: string;
}

File: api/v1/schemas/btc_training.py
# -*- coding: utf-8 -*-
"""Schemas for the offline BTC model research workspace."""

from __future__ import annotations

from typing import Any, Dict, List, Literal, Optional

from pydantic import BaseModel, Field


class BtcTrainingRunRequest(BaseModel):
    architecture: Literal["patchtst", "itransformer", "fusion", "all"] = "patchtst"
    epochs: int = Field(30, ge=20, le=200, description="Research runs default to 30 epochs; at least 20 are required")
    seeds: List[int] = Field(
        default_factory=lambda: [7, 13, 29, 43, 71],
        min_length=5,
        max_length=20,
        description="At least five distinct seeds are required before judging a structure",
    )
    sequence_length: int = Field(256, ge=8, le=2048)
    folds: int = Field(12, ge=1, le=50)
    min_train_samples: int = Field(5000, ge=24, le=200000)
    validation_samples: int = Field(168, ge=1, le=20000)
    purge_samples: int = Field(48, ge=1, le=20000)
    ablation_features: List[str] = Field(default_factory=list, max_length=128)


class BtcTrainingTaskAccepted(BaseModel):
    task_id: str
    status: str
    message: Optional[str] = None
    protocol: Dict[str, Any] = Field(default_factory=dict)


class BtcTrainingTaskStatus(BaseModel):
    task_id: str
    status: str
    progress: int = 0
    message: Optional[str] = None
    result: Optional[Dict[str, Any]] = None
    error: Optional[str] = None


class BtcTrainingConfigResponse(BaseModel):
    architectures: List[str]
    default_seeds: List[int]
    min_seeds: int
    min_epochs: int
    feature_count: int
    feature_columns: List[str]
    research_only: bool = True
    promotion_eligible: bool = False
